radicle/src/storage/git/transport/local/url.rs

```diff
@@ -0,0 +1,106 @@
+//! Git local transport URLs.
+use std::str::FromStr;
+
+use thiserror::Error;
+
+use crate::{
+    crypto,
+    identity::{Id, IdError},
+};
+
+/// Repository namespace.
+type Namespace = crypto::PublicKey;
+
+#[derive(Debug, Error)]
+pub enum UrlError {
+    /// Invalid format.
+    #[error("invalid url format: expected `rad://<repo>[/<namespace>]`")]
+    InvalidFormat,
+    /// Unsupported URL scheme.
+    #[error("unsupported scheme: expected `rad://`")]
+    UnsupportedScheme,
+    /// Invalid repository identifier.
+    #[error("repo: {0}")]
+    InvalidRepository(#[source] IdError),
+    /// Invalid namespace.
+    #[error("namespace: {0}")]
+    InvalidNamespace(#[source] crypto::PublicKeyError),
+}
```

Comment:
Is it rust idiom to prepend the error with the type, even in cases where it has a dedicated module?

Reply:
Tbh, I avoid it. I usually use `Error` if it's a single error in the module or I have an `error` module with different classes of errors:
```rust
// storage/error.rs

/// Things go wrong during `find`
pub enum Find {
  /* variants */
}

/// Things go wrong during `update`
pub enum Update {
  /* variants */
}
```

I think you'll find all sorts of ways in Rust, including the dreaded God `Error` -- it has all the error variants and _all_ your functions return it so you don't know how they particularly fail it could be any of the variants really.

Reply:
Yeah they are both idiomatic. I tend to use prefixes when the modules are re-exported; eg. in this case I'm accessing the types via `transport::remote::{Url, UrlError}`, everywhere, not `transport::remote::url::Url`.